src/shared/pkcs11-util.h

```diff
@@ -1,6 +1,9 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 #pragma once
 
+#if HAVE_OPENSSL
+#  include <openssl/x509.h>
+#endif
```

Comment:
This looks iffy. The include here is conditionalized on HAVE_OPENSSL, and the function signature below on HAVE_P11KIT. Either the include is not necessary, or this will fail to build in some combinations.

Reply:
Don't understand what's wrong here. Could you provide an example when it fails to build?

Reply:
I don't have an example. I'm just following simple logic: an include should be added to a header iff this include is required by something later in that header. If that later something is conditionalized differently than the include of the header, then normally we would expect things to break when the later something is defined, but the header is not included.

But OK, let's leave this as is. We will know soon enough if it causes any compilation issues and is should be trivial to fix.

Reply:
I think you are missing that functions that use `openssl/x509.h` are also guarded by `#if HAVE_OPENSSL`:

```
#if HAVE_OPENSSL
#  include <openssl/x509.h>
#endif
<...>

#if HAVE_P11KIT
<...>
#if HAVE_OPENSSL
int pkcs11_token_read_x509_certificate(CK_FUNCTION_LIST *m, CK_SESSION_HANDLE session, CK_OBJECT_HANDLE object, X509 **ret_cert);
#endif
<...>
#if HAVE_OPENSSL
int pkcs11_acquire_certificate(const char *uri, const char *askpw_friendly_name, const char *askpw_icon_name, X509 **ret_cert, char **ret_pin_used);
#endif
<...>
#else
static inline int dlopen_p11kit(void) {
        return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "p11kit support is not compiled in.");
}
#endif
```
So the maximum issue I see here is including an extra header when `HAVE_OPENSSL == 1` and `HAVE_P11KIT == 0`.